### PR TITLE
[WIP] canisters will rupture or explode when overpressured

### PR DESCRIPTION
**Pull request by hornygranny**
the define values are placeholders, the proc is lifted from tanks but is being rewritten locally

### CODE REVIEW

**Comment by WJohn** [comment]
So uh, how exactly would one go about making bombs, then?

**Comment by hornygranny** [comment]
the threshold is pretty high, burning 66/33 mix only started leaking from a canister at 15000 kPa

**Comment by Razharas** [comment]
>I cant collect all plasma and oxygen in atmos in one canister to open up on asteroid to devoid it of life for the rest of the round anymore
I never seen a hitler and hg in the same room my entire life
Coincidence?
I THINK NOT

**Comment by octareenroon91** [comment]
Who are you even >implying? Why would a hitler (sic) make his work harder to do?

To say nothing of bringing several canisters and a pipe dispenser to the asteroid...

**Comment by neersighted** [comment]
Canisters should really process_atmos...

**Comment by neersighted** [comment]
Also a color indicator warning at near leak and rupture pressures would be nice.

**Comment by CPTANT** [comment]
canisters do process atmos.

Anyway, this seems like a logical change.

**Comment by neersighted** [comment]
I mean for their main process, like this bursting. Anyway, looking pretty good.

**Comment by Fox-McCloud** [comment]
Canister bombs, when?

**Comment by Buggy123** [comment]
I'll support this change if, and only if, we get massive canister bombs in return.

**Comment by Robertha89** [comment]
isnt the canister limit like 2.5 million kpa?

**Comment by Niknakflak** [comment]
This removes internal canister combustion?  I'm a big fan of internal gas mixing and reacting to get a canister at like 999999999999999kpa and temperature of 99999999999k and then using heat transfer on a more pure canister for maximum potential.  Personally a :-1: but not a deal breaker for me tbh

**Comment by octareenroon91** [comment]
This code implies canister bombs.

Going with what Niknakflak says, generous limits will probably make this change extra popular.

... what if the containment concept were generalized to all gas_mixtures?

File: code/__DEFINES/atmospherics.dm [hunk 1]
```diff
@@ -30,6 +30,10 @@
 #define TANK_RUPTURE_PRESSURE				(40.*ONE_ATMOSPHERE)	//Tank spills all contents into atmosphere
 #define TANK_FRAGMENT_PRESSURE				(50.*ONE_ATMOSPHERE)	//Boom 3x3 base explosion
 #define TANK_FRAGMENT_SCALE	    			(10.*ONE_ATMOSPHERE)	//+1 for each SCALE kPa aboe threshold
+#define CANISTER_LEAK_PRESSURE					(150.*ONE_ATMOSPHERE)	//Tank starts leaking
+#define CANISTER_RUPTURE_PRESSURE				(200.*ONE_ATMOSPHERE)	//Tank spills all contents into atmosphere
+#define CANISTER_FRAGMENT_PRESSURE				(250.*ONE_ATMOSPHERE)	//Boom 3x3 base explosion
+#define CANISTER_FRAGMENT_SCALE	    			(50.*ONE_ATMOSPHERE)	//+1 for each SCALE kPa aboe threshold
```
Comment (neersighted) [inline code/__DEFINES/atmospherics.dm]: grrr indentation

File: code/game/machinery/atmoalter/canister.dm [hunk 2]
```diff
@@ -367,3 +369,53 @@ update_flag
...
+/obj/machinery/portable_atmospherics/canister/proc/check_status()
+	//Handle exploding, leaking, and rupturing of the tank
+
+	if(!air_contents)
+		return 0
+
+	var/pressure = air_contents.return_pressure()
+	if(pressure > CANISTER_FRAGMENT_PRESSURE)
+		if(!istype(src.loc,/obj/item/device/transfer_valve))
+			message_admins("Explosive tank rupture! Last key to touch the tank was [src.fingerprintslast].")
```
Comment (neersighted) [inline code/game/machinery/atmoalter/canister.dm]: canister

File: code/game/machinery/atmoalter/canister.dm [hunk 3]
```diff
@@ -367,3 +369,53 @@ update_flag
...
+			log_game("Explosive tank rupture! Last key to touch the tank was [src.fingerprintslast].")
+		world << "\blue[x],[y] tank is exploding: [pressure] kPa"
+		//Give the gas a chance to build up more pressure through reacting
+		air_contents.react()
+		air_contents.react()
```
Comment (neersighted) [inline code/game/machinery/atmoalter/canister.dm]: Please loop or do something less shitty here.

File: code/game/machinery/atmoalter/canister.dm [hunk 4]
```diff
@@ -367,3 +369,53 @@ update_flag
...
+		air_contents.react()
+		pressure = air_contents.return_pressure()
+		var/range = (pressure-CANISTER_FRAGMENT_PRESSURE)/CANISTER_FRAGMENT_SCALE
```
Comment (neersighted) [inline code/game/machinery/atmoalter/canister.dm]: spaces for readability

File: code/game/machinery/atmoalter/canister.dm [hunk 5]
```diff
@@ -367,3 +369,53 @@ update_flag
...
+		var/turf/epicenter = get_turf(loc)
+
+		world << "\blue Exploding Pressure: [pressure] kPa, intensity: [range]"
+
+		explosion(epicenter, round(range*0.25), round(range*0.5), round(range), round(range*1.5))
```
Comment (neersighted) [inline code/game/machinery/atmoalter/canister.dm]: spaces again

File: code/game/machinery/atmoalter/canister.dm [hunk 6]
```diff
@@ -367,3 +369,53 @@ update_flag
...
+		if(istype(src.loc,/obj/item/device/transfer_valve))
```
Comment (neersighted) [inline code/game/machinery/atmoalter/canister.dm]: This makes no sense for canisters.

**Comment by KazeEspada** [comment]
@neersighted Its just a bunch of Copypasta at the moment, he needs to go and fix it first.

**Comment by hornygranny** [comment]
most of that is already fixed locally

**Comment by KazeEspada** [comment]
Yeah, I wasn't going to hold it against you. I expected you to fix it before the WIP was removed.

File: code/game/machinery/atmoalter/canister.dm (hunk 3 above)
Comment (Fox-McCloud) [inline code/game/machinery/atmoalter/canister.dm]: Not saying it's justified, but this is how tank transfer bombs work.

**Comment by duncathan** [comment]
This is shitty copy paste
Generalize it to any container or I'm gonna say no

**Comment by Lumipharon** [comment]
Am I missing something here? Canister bombs are already a thing, unless something has changed, so the ony other way to bloat a canister other than combustion I presume would be a volumetric pump and a shitload of gas?

**Comment by Fox-McCloud** [comment]
@Lumipharon 

Canister bombs are a thing where you rig up a canister with an assembly to blow it; the potential explosion is several orders bigger than tank transfer valves, but actually utilizing one effectively is difficult.

File: code/game/machinery/atmoalter/canister.dm [hunk 7]
```diff
@@ -367,3 +369,46 @@ update_flag
...
+
+/obj/machinery/portable_atmospherics/canister/proc/check_status()
+	//Handle exploding, leaking, and rupturing of the tank
+
+	if(!air_contents)
+		return 0
+
+	var/pressure = air_contents.return_pressure()
+	if(pressure > CANISTER_FRAGMENT_PRESSURE)
+		if(!istype(src.loc,/obj/item/device/transfer_valve))
```
Comment (duncathan) [inline code/game/machinery/atmoalter/canister.dm]: This check is unnecessary

**Comment by Lumipharon** [comment]
@Fox-McCloud I know what cannister bombs are, I've used them before. 
I'm confused by comments here that suggest this pr is making canister bombs possible when they already are.

File: code/game/machinery/atmoalter/canister.dm [hunk 8]
```diff
@@ -367,3 +369,46 @@ update_flag
...
+			message_admins("Explosive canister rupture! Last key to touch the tank was [src.fingerprintslast].")
+			log_game("Explosive canister rupture! Last key to touch the tank was [src.fingerprintslast].")
+		//world << "\blue[x],[y] tank is exploding: [pressure] kPa"
+		pressure = air_contents.return_pressure()
+		var/range = (pressure-CANISTER_FRAGMENT_PRESSURE)/CANISTER_FRAGMENT_SCALE
```
Comment (neersighted) [inline code/game/machinery/atmoalter/canister.dm]: space operators please

File: code/game/machinery/atmoalter/canister.dm [hunk 9]
```diff
@@ -367,3 +369,46 @@ update_flag
...
+		//world << "\blue[x],[y] canister is rupturing: [pressure] kPa, integrity [integrity]"
+		if(integrity <= 0)
+			health = 0
+			healthcheck()
+		else
+			integrity--
+
+	else if(pressure > CANISTER_LEAK_PRESSURE)
+		//world << "\blue[x],[y] canister is leaking: [pressure] kPa, integrity [integrity]"
+		if(integrity <= 0)
```
Comment (neersighted) [inline code/game/machinery/atmoalter/canister.dm]: This can just if `if(integrity-- <=0)`.

File: code/game/machinery/atmoalter/canister.dm [hunk 10]
```diff
@@ -193,6 +194,7 @@ update_flag
 
 /obj/machinery/portable_atmospherics/canister/process()
 	src.updateDialog()
+	check_status()
```
Comment (neersighted) [inline code/game/machinery/atmoalter/canister.dm]: a better proc name would be nice. also explosions and atmos updates happen in this proc, it should be in process_atmos.

File: code/game/machinery/atmoalter/canister.dm [hunk 11]
```diff
@@ -193,6 +194,7 @@ update_flag
 
 /obj/machinery/portable_atmospherics/canister/process()
 	src.updateDialog()
```
Comment (neersighted) [inline code/game/machinery/atmoalter/canister.dm]: Oops, tgui handles all of this nonsense. Please remove all calls to updateDialog/updateUsrDialog.

File: code/game/machinery/atmoalter/canister.dm [hunk 12]
```diff
@@ -367,3 +368,42 @@ update_flag
...
+		//world << "\blue[x],[y] canister is leaking: [pressure] kPa, integrity [integrity]"
+		if(integrity-- <= 0)
+			var/turf/simulated/T = get_turf(src)
+			if(!T)
+				return
+			var/datum/gas_mixture/leaked_gas = air_contents.remove_ratio(0.25)
+			T.assume_air(leaked_gas)
+
+	else if(integrity < 3)
+		integrity++
```
Comment (duncathan) [inline code/game/machinery/atmoalter/canister.dm]: These two lines could be replaced with `integrity = min(++integrity, 3)` but it's really not a big deal

File: code/game/machinery/atmoalter/canister.dm [hunk 13]
```diff
@@ -192,7 +193,7 @@ update_flag
 			update_icon()
 
 /obj/machinery/portable_atmospherics/canister/process()
-	src.updateDialog()
+	check_overpressure()
```
Comment (neersighted) [inline code/game/machinery/atmoalter/canister.dm]: Should still really be in process_atmos

File: code/game/machinery/atmoalter/canister.dm [hunk 14]
```diff
@@ -367,3 +368,42 @@ update_flag
...
+			health = 0
+			healthcheck()
+
+	else if(pressure > CANISTER_LEAK_PRESSURE)
+		//world << "\blue[x],[y] canister is leaking: [pressure] kPa, integrity [integrity]"
+		if(integrity-- <= 0)
+			var/turf/simulated/T = get_turf(src)
+			if(!T)
+				return
+			var/datum/gas_mixture/leaked_gas = air_contents.remove_ratio(0.25)
```
Comment (neersighted) [inline code/game/machinery/atmoalter/canister.dm]: Maybe base this on integrity incase someone raises/lowers it.
`1 / (integerity + 1)`

**Comment by tkdrg** [comment]
@hornygranny are you still working on this

**Comment by neersighted** [comment]
@tkdrg I've been meaning to refactor canisters, I'll port the code to be generic to both.

**Pull request closed without merging**